### PR TITLE
openapi3: reference originating locations in YAML specs - step 1

### PR DESCRIPTION
**Pull request by reuvenharrison**
This PR adds the ability to correlate OpenAPI components with the location in the originating YAML spec.
Here's a summary of the changes required to achieve this:
1. Modify [yaml.v3 ](https://github.com/go-yaml/yaml/tree/v3) to add line and column numbers of each YAML element to the unmarshaled interfaces (see https://github.com/oasdiff/yaml3)
2. Modify [invopop/yaml](https://github.com/invopop/yaml) with a new function: `UnmarshalWithOrigin` (see https://github.com/oasdiff/yaml)
3. Modify this repo to add an `Origin` struct to OpenAPI components. This requires changes to the UnmarshalJSON functions and, in some cases, replacing native types, like `map[string]string`, by dedicated types that include the original native type and an Origin member.

This PR addresses issue https://github.com/getkin/kin-openapi/issues/986

Follow up tasks to be addressed by future PRs:
- Add `origin` to more elements, namely `Components` and some types that are unmarshalled with `unmarshalStringMapP` or `unmarshalStringMap`
- Merge changes in yaml.v3 and invopop/yaml back to original repos and remove `replace`  directives from go.mod
- Enhance `origin` to include end line and column and YAML comments as proposed by @fenollp

### CODE REVIEW

File: openapi3/components.go [hunk 1]
```diff
@@ -110,6 +112,60 @@ func (components *Components) UnmarshalJSON(data []byte) error {
...
+func (schemas *Schemas) UnmarshalJSON(data []byte) (err error) {
+	*schemas, _, err = unmarshalStringMapP[SchemaRef](data)
+	return
+}
+
+// UnmarshalJSON sets SecuritySchemes to a copy of data.
+func (securitySchemes *SecuritySchemes) UnmarshalJSON(data []byte) (err error) {
+	*securitySchemes, _, err = unmarshalStringMapP[SecuritySchemeRef](data)
+	return
+}
```
Comment (fenollp) [inline openapi3/components.go:167]: Please define these in their respective files

File: openapi3/discriminator.go [hunk 2]
```diff
@@ -9,9 +9,10 @@ import (
 // See https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.0.3.md#discriminator-object
 type Discriminator struct {
 	Extensions map[string]any `json:"-" yaml:"-"`
+	Origin     *Origin        `json:"origin,omitempty" yaml:"origin,omitempty"`
 
-	PropertyName string            `json:"propertyName" yaml:"propertyName"` // required
-	Mapping      map[string]string `json:"mapping,omitempty" yaml:"mapping,omitempty"`
+	PropertyName string    `json:"propertyName" yaml:"propertyName"` // required
+	Mapping      StringMap `json:"mapping,omitempty" yaml:"mapping,omitempty"`
```
Comment (fenollp) [inline openapi3/discriminator.go:15]: Please move the introduction of the `StringMap` type to another PR, with a breaking change note to README if applicable.

File: openapi3/security_scheme.go [hunk 3]
```diff
@@ -316,11 +320,12 @@ func (flows *OAuthFlows) Validate(ctx context.Context, opts ...ValidationOption)
...
+	Origin     *Origin        `json:"origin,omitempty" yaml:"origin,omitempty"`
 
-	AuthorizationURL string            `json:"authorizationUrl,omitempty" yaml:"authorizationUrl,omitempty"`
-	TokenURL         string            `json:"tokenUrl,omitempty" yaml:"tokenUrl,omitempty"`
-	RefreshURL       string            `json:"refreshUrl,omitempty" yaml:"refreshUrl,omitempty"`
-	Scopes           map[string]string `json:"scopes" yaml:"scopes"` // required
+	AuthorizationURL string    `json:"authorizationUrl,omitempty" yaml:"authorizationUrl,omitempty"`
+	TokenURL         string    `json:"tokenUrl,omitempty" yaml:"tokenUrl,omitempty"`
+	RefreshURL       string    `json:"refreshUrl,omitempty" yaml:"refreshUrl,omitempty"`
+	Scopes           StringMap `json:"scopes" yaml:"scopes"` // required
```
Comment (fenollp) [inline openapi3/security_scheme.go:328]: ditto

**Review by fenollp — CHANGES_REQUESTED** [review]
A few minor edits otherwise LGTM

File: openapi3/components.go (hunk 1 above)
Comment (reuvenharrison) [reply openapi3/components.go:167]: Done

File: openapi3/discriminator.go (hunk 2 above)
Comment (reuvenharrison) [reply openapi3/discriminator.go:15]: Done: https://github.com/getkin/kin-openapi/pull/1018
After https://github.com/getkin/kin-openapi/pull/1018 is merged, I will update this one to leverage it.

File: openapi3/security_scheme.go (hunk 3 above)
Comment (reuvenharrison) [reply openapi3/security_scheme.go:328]: Done

File: openapi3/discriminator.go (hunk 2 above)
Comment (reuvenharrison) [reply openapi3/discriminator.go]: Yes, I was struggling but managed to do it eventually. Thanks.

**Review by fenollp — APPROVED** [review]

**Pull request merged**